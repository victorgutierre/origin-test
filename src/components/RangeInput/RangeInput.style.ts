import styled from 'styled-components';

export const RangeInputWrapper = styled.div`
  width: 40%;
`;

export const Label = styled.p`
  display: block;
  margin-bottom: 0.5rem;
  color: #202a31;
  font-size: 1rem;
  text-align: left;
`;

export const RangeInputBox = styled.div`
  display: flex;
  justify-content: space-between;
  padding: 1rem;
  border-radius: 8px;
  border: 1px solid #eaeef2;
  position: relative;
`;

export const Text = styled.p`
  color: red;
  margin: 0 1rem;
  text-align: center;
  vertical-align: middle;
  color: #748695;

  b {
    display: block;
    color: #000;
    margin-bottom: 0.5rem;
  }
`;

export const Decrease = styled.div`
  cursor: pointer;

  &:before,
  &:after {
    content: '';
    display: block;
    height: 1rem;
    margin-top: -13px;
    position: absolute;
    transform: rotate(-135deg);
    left: 1rem;
    top: 50%;
    width: 0;
    border-right: 2px solid #8d9ca8;
  }

  &:after {
    margin-top: -3px;
    transform: rotate(-45deg);
  }
`;

export const Increase = styled.div`
  cursor: pointer;

  &:before,
  &:after {
    content: '';
    display: block;
    height: 1rem;
    margin-top: -13px;
    position: absolute;
    transform: rotate(135deg);
    right: 1rem;
    top: 50%;
    width: 0;
    border-right: 2px solid #8d9ca8;
  }

  &:after {
    margin-top: -3px;
    transform: rotate(45deg);
  }
`;
