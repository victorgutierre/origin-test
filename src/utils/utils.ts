export const months = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December'
];

export const money = (str: string) => {
  if (!str) return '';

  return parseFloat(str)
    .toFixed(2)
    .replace(/(\d)(?=(\d{3})+\.)/g, '$1,')
    .replace(/.(\d{2})$/g, '.$1');
}
