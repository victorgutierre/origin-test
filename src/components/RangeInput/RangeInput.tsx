import React from 'react';
import { months } from './../../utils/months';
import {
  RangeInputWrapper,
  Label,
  RangeInputBox,
  Text,
  Decrease,
  Increase
} from './RangeInput.style';
interface Props {
  actualMonth: number;
  actualYear: number;
  month: number;
  year: number;
  differenceYearsInMonths: number;
  setMonth: (arg: number) => void;
  setYear: (arg: number) => void;
  setYearsInMonths: (arg: number) => void;
}

export const RangeInput: React.FC<Props> = ({
  actualMonth,
  actualYear,
  month,
  year,
  differenceYearsInMonths,
  setMonth,
  setYear,
  setYearsInMonths
}) => {
  const increase = () => {
    if (month === 11) {
      setYear(year + 1);
      setMonth(0);
    } else {
      setMonth(month + 1);
    }

    setYearsInMonths(differenceYearsInMonths + 1);
  };

  const decrease = () => {
    if (month === 0) {
      setYear(year - 1);
      setMonth(11);
    } else {
      setMonth(month - 1);
    }

    setYearsInMonths(differenceYearsInMonths - 1);
  };

  const disableButton = () => month === actualMonth && year === actualYear;

  return (
    <RangeInputWrapper>
      <Label>Reach goal by</Label>

      <RangeInputBox>
        <Decrease onClick={() => decrease()} disabled={disableButton()} />
        <Text>
          <b>{months[month]}</b> {year}
        </Text>
        <Increase onClick={() => increase()} />
      </RangeInputBox>
    </RangeInputWrapper>
  );
};
