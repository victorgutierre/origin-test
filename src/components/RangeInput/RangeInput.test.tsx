import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import { fireEvent, queryByTestId, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { RangeInput } from './RangeInput';
import { months } from '../../utils/utils';

describe('<RangeInput />', () => {
  const actualDate = new Date();
  const actualFullYear = actualDate.getFullYear();
  const actualMonth = actualDate.getMonth();
  const defaultYearPlanning = 4;

  const differenceYearsInMonths = defaultYearPlanning * 12;
  const month = actualMonth;
  const year = actualFullYear + defaultYearPlanning;
  const yearsInMonths = differenceYearsInMonths;

  const setMonth = jest.fn();
  const setYear = jest.fn();
  const setYearsInMonths = jest.fn();

  const component = (
    <RangeInput
      actualYear={actualFullYear}
      actualMonth={actualMonth}
      month={month}
      year={year}
      differenceYearsInMonths={yearsInMonths}
      setMonth={setMonth}
      setYear={setYear}
      setYearsInMonths={setYearsInMonths}
    />
  );

  it('Should render the RangeInput component', () => {
    const wrapper = shallow(component);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Month element must render', () => {
    const { queryByTestId } = render(component);
    const month = queryByTestId('range-input-month');
    expect(month).toBeTruthy();
  });

  it('Month text must be same as converted', () => {
    const { queryByTestId } = render(component);
    const monthText = queryByTestId('range-input-month').textContent;
    expect(monthText).toBe(months[month]);
  });

  it('Year element must render', () => {
    const { queryByTestId } = render(component);
    const year = queryByTestId('range-input-year');
    expect(year).toBeTruthy();
  });

  it('Year text must be 2025', () => {
    const { queryByTestId } = render(component);
    const yearText = queryByTestId('range-input-year').textContent;
    expect(yearText).toBe('2025');
  });

  it('Click 1 time increase should call month and setYearsInMonths func', () => {
    const { queryByTestId } = render(component);
    const element = queryByTestId('range-input-increase');
    fireEvent.click(element);

    expect(setMonth).toHaveBeenCalled();
    expect(setYear).not.toHaveBeenCalled();
    expect(setYearsInMonths).toHaveBeenCalled();
  });

  it('Click 1 time decrease should call month and setYearsInMonths func', () => {
    const { queryByTestId } = render(component);
    const element = queryByTestId('range-input-decrease');
    fireEvent.click(element);

    expect(setMonth).toHaveBeenCalled();
    expect(setYear).not.toHaveBeenCalled();
    expect(setYearsInMonths).toHaveBeenCalled();
  });

  it('Click 1 time increase button value should be 4', () => {
    const { queryByTestId } = render(component);
    const element = queryByTestId('range-input-increase');
    fireEvent.click(element, { target: { value: month + 1 } });

    expect(element.getAttribute('value')).toBe('4');
  });

  it('Click 1 time decrease button value should be 2', () => {
    const { queryByTestId } = render(component);
    const element = queryByTestId('range-input-decrease');
    fireEvent.click(element, { target: { value: month - 1 } });

    expect(element.getAttribute('value')).toBe('2');
  });
});

describe('<RangeInput /> when month is current month', () => {
  const actualDate = new Date();
  const actualFullYear = actualDate.getFullYear();
  const actualMonth = actualDate.getMonth();

  const yearsInMonths = 4 * 12;

  const setMonth = jest.fn();
  const setYear = jest.fn();
  const setYearsInMonths = jest.fn();

  const component = (
    <RangeInput
      actualYear={actualFullYear}
      actualMonth={actualMonth}
      month={actualMonth}
      year={actualFullYear}
      differenceYearsInMonths={yearsInMonths}
      setMonth={setMonth}
      setYear={setYear}
      setYearsInMonths={setYearsInMonths}
    />
  );

  it('Decrease element should be disabled', () => {
    const { queryByTestId } = render(component);
    const element = queryByTestId('range-input-decrease');
    expect(element.getAttribute('disabled')).toBe('');
  });
});

describe('<RangeInput /> when month is december (11)', () => {
  const actualDate = new Date();
  const actualFullYear = actualDate.getFullYear();
  const actualMonth = actualDate.getMonth();

  const yearsInMonths = 4 * 12;

  const setMonth = jest.fn();
  const setYear = jest.fn();
  const setYearsInMonths = jest.fn();

  const component = (
    <RangeInput
      actualYear={actualFullYear}
      actualMonth={actualMonth}
      month={11}
      year={actualFullYear}
      differenceYearsInMonths={yearsInMonths}
      setMonth={setMonth}
      setYear={setYear}
      setYearsInMonths={setYearsInMonths}
    />
  );

  it('Click 1 time increase button when month is december value should be 0', () => {
    const { queryByTestId } = render(component);
    const element = queryByTestId('range-input-increase');
    fireEvent.click(element, { target: { value: 0 } });

    expect(element.getAttribute('value')).toBe('0');
  });
});

describe('<RangeInput /> when month is january (0)', () => {
  const actualDate = new Date();
  const actualFullYear = actualDate.getFullYear();
  const actualMonth = actualDate.getMonth();

  const yearsInMonths = 4 * 12;

  const setMonth = jest.fn();
  const setYear = jest.fn();
  const setYearsInMonths = jest.fn();

  const component = (
    <RangeInput
      actualYear={actualFullYear}
      actualMonth={actualMonth}
      month={0}
      year={actualFullYear}
      differenceYearsInMonths={yearsInMonths}
      setMonth={setMonth}
      setYear={setYear}
      setYearsInMonths={setYearsInMonths}
    />
  );

  it('Click 1 time decrease button when month is december value should be 11', () => {
    const { queryByTestId } = render(component);
    const element = queryByTestId('range-input-decrease');
    fireEvent.click(element, { target: { value: 11 } });

    expect(element.getAttribute('value')).toBe('11');
  });
});
