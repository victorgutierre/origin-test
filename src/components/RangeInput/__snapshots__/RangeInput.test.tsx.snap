// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RangeInput /> Should render the RangeInput component 1`] = `
<styled.div>
  <styled.p>
    Reach goal by
  </styled.p>
  <styled.div>
    <styled.button
      data-testid="range-input-decrease"
      disabled={false}
      onClick={[Function]}
      value={3}
    />
    <styled.p>
      <b
        data-testid="range-input-month"
      >
        April
      </b>
       
      <span
        data-testid="range-input-year"
      >
        2025
      </span>
    </styled.p>
    <Styled(styled.button)
      data-testid="range-input-increase"
      onClick={[Function]}
      value={3}
    />
  </styled.div>
</styled.div>
`;
